internal/serviceconfig/duration.go

```diff
@@ -0,0 +1,117 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package serviceconfig
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// Duration defines JSON marshal and unmarshal methods to conform to the
+// protobuf JSON spec defined [here].
+//
+// [here]: https://protobuf.dev/reference/protobuf/google.protobuf/#duration
+type Duration time.Duration
+
+func (d Duration) String() string {
+	return fmt.Sprint(time.Duration(d))
+}
+
+// MarshalJSON converts from d to a JSON string output.
+func (d Duration) MarshalJSON() ([]byte, error) {
+	ns := time.Duration(d).Nanoseconds()
+	sec := ns / int64(time.Second)
+	ns = ns % int64(time.Second)
+
+	var sign string
+	if sec < 0 {
+		sign, sec, ns = "-", -1*sec, -1*ns
+	}
+
+	// Generated output always contains 0, 3, 6, or 9 fractional digits,
+	// depending on required precision.
+	str := fmt.Sprintf("%s%d.%09d", sign, sec, ns)
+	str = strings.TrimSuffix(str, "000")
+	str = strings.TrimSuffix(str, "000")
+	str = strings.TrimSuffix(str, ".000")
+	return []byte(fmt.Sprintf("\"%ss\"", str)), nil
+}
+
+// UnmarshalJSON unmarshals b as a duration JSON string into d.
+func (d *Duration) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+	if !strings.HasSuffix(s, "s") {
+		return fmt.Errorf("malformed duration %q", s)
+	}
+	ss := strings.SplitN(s[:len(s)-1], ".", 3)
+	if len(ss) > 2 {
+		return fmt.Errorf("malformed duration %q", s)
+	}
+	// hasDigits is set if either the whole or fractional part of the number is
+	// present, since both are optional but one is required.
+	hasDigits := false
+	if len(ss[0]) > 0 {
+		sec, err := strconv.ParseInt(ss[0], 10, 64)
+		if err != nil {
+			return fmt.Errorf("malformed duration %q: %v", s, err)
+		}
+		const maxSeconds = math.MaxInt64 / int64(time.Second)
+		const minSeconds = math.MinInt64 / int64(time.Second)
+		if sec > maxSeconds || sec < minSeconds {
+			return fmt.Errorf("out of range: %q", s)
+		}
+		*d = Duration(sec) * Duration(time.Second)
+		hasDigits = true
+	} else {
+		*d = 0
+	}
+	if len(ss) == 2 && len(ss[1]) > 0 {
+		if len(ss[1]) > 9 {
+			return fmt.Errorf("malformed duration %q", s)
+		}
+		f, err := strconv.ParseInt(ss[1], 10, 64)
+		if err != nil {
+			return fmt.Errorf("malformed duration %q: %v", s, err)
+		}
+		neg := false
+		if *d < 0 {
+			neg = true
+			f *= -1
+		}
+		for i := 9; i > len(ss[1]); i-- {
+			f *= 10
+		}
+		*d += Duration(f)
+		if neg != (*d < 0) {
+			return fmt.Errorf("out of range: %q", s)
+		}
+		hasDigits = true
```

Comment:
I don't think flipping the bit here is the only thing can determine out of range? The language states this "Durations less than one second are represented with a 0 seconds field and a positive or negative nanos field.  For durations of one second or more, a non-zero value for the nanos field must be of the same sign as the seconds field." So if total duration is greater than 1 sec, you need to check if the signs are the same" If sec is 0 and nanoseconds is set, what will happen with this check? Seems like a valid case.

Reply:
> "Durations less than one second are represented with a 0 seconds field and a positive or negative nanos field. For durations of one second or more, a non-zero value for the nanos field must be of the same sign as the seconds field."

I don't understand why you're looking at the protobuf Duration representation at all.  We don't use that.  Go just uses a signed int64 to store durations, and we're converting from a string.  I'm not sure how else you'd get the wrong sign at this point, since we check the seconds field against max/min int64s (after multiplying by 1B (AKA time.Second)) to ensure they're within range.  This catches the case where seconds is == max/min int64 * 1B and the nanoseconds causes us to wrap.  In this case we will flip the sign bit.  There are test cases for both immediately below and immediately above the overflow threshold in both positive and negative cases.

Reply:
> Plus this line: "Must be from -999,999,999 to +999,999,999 inclusive."

There's a length check in there of 9 digits.

Reply:
What? How could I not look at the protobuf duration representation. Or are you saying it's a higher order of logic? -999,999,999 to +999,999,999 this is within bounds of the unit64.